KAFKA-12309 The revocation algorithm produces uneven distributions

chia7712 opened the pull request:
issue: https://issues.apache.org/jira/browse/KAFKA-12309

# issue description
When adding a new worker, the revocation algorithm tries to revoke some connectors/tasks from existent workers so as to move them to new worker to balance load. However, the algorithm can revoke incorrect number of connectors/tasks for following cases.


### Assignments:
"worker_0" -> 8 connectors/tasks
"worker_1" -> 8 connectors/tasks
(New) "worker_2" -> 0 connectors/tasks

### Result
"worker_0" -> 6 connectors/tasks
"worker_1" -> 6 connectors/tasks
"worker_2" -> 4 connectors/tasks

### (expected) Result
"worker_0" -> 6 connectors/tasks
"worker_1" -> 5 connectors/tasks
"worker_2" -> 5 connectors/tasks
or 
"worker_0" -> 5 connectors/tasks
"worker_1" -> 6 connectors/tasks
"worker_2" -> 5 connectors/tasks

# root cause
the algorithm makes all existent workers keep `ceil` (`numToRevoke = existing.tasksSize() - ceilTasks;`) connectors/tasks. That results in that there is no enough revoked connectors/tasks to be assigned to new worker.

# extra changes included by this PR

1. move revocation algorithm to a helper method in order to test it


### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

chia7712 committed:
KAFKA-12309 The revocation algorithm produces uneven distributions

chia7712 commented on connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java:
```diff
@@ -359,7 +359,9 @@ private boolean assertConnectorAndTasksAreUniqueAndBalanced() {
                     tasks.values().size(),
                     tasks.values().stream().distinct().collect(Collectors.toList()).size());
             assertTrue("Connectors are imbalanced: " + formatAssignment(connectors), maxConnectors - minConnectors < 2);
+            if (minConnectors > 1) assertEquals("Some workers have no connectors", connectors.size(), connect.workers().size());
```
make sure there is no idle worker

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 commented:
@rhauch Could you take a look? this uneven distributions can be reproduced easily so it would be nice to fix it asap.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
-            for (int i = existing.tasksSize(); i > floorTasks && numToRevoke > 0; --i, --numToRevoke) {
+            int currentSize = forTask ? existing.tasksSize() : existing.connectorsSize();
+            int expectedSize;
+            if (existingWorkers.size() == 1 || currentSize == 1) expectedSize = ceil;
```
Why do we need this condition? And why under this case, we don't need to have `numberOfBiggers--`? After all, we assigned one ceiling to this worker, didn't we?

chia7712 replied:
That is a specify case as the node having single task is already in balance. We can move the job to another node but it does not balance anything but it brings extra down-time.

I will add comments for that case.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
+    static void computeRevoked(Map<String, ConnectorsAndTasks> revoking,
+                               Collection<WorkerLoad> existingWorkers,
+                               int numberOfTotalWorks,
+                               int numberOfActives,
```
Could we rename this parameters as `numberOfActiveTasks` to be more clear?

showuon replied:
And they can be `final`

chia7712 replied:
nice catch. will copy that.

kkonstantine replied:
Let's not change local variables and arguments with `final` when there isn't this style in place. Sorry for the back and forth.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -562,50 +561,63 @@ private void resetDelay() {
+                               int numberOfActives,
+                               boolean forTask) {
+        int floor = numberOfActives / numberOfTotalWorks;
+        int numberOfBiggers = numberOfActives % numberOfTotalWorks;
```
I believe you're trying to name it `numberOfRemainder`

showuon replied:
Oh, I got what you mean, maybe `numberOfCeilingMembers` would be more clear?

chia7712 replied:
ok. will copy that

showuon reviewed (commented):
@chia7712 , thanks for the patch! This is a good finding! Left some comments. Thanks.

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
make better naming

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java:
```diff
@@ -562,50 +561,68 @@ private void resetDelay() {
+                               final int numberOfWorkers,
+                               final int numberOfActiveTasks,
+                               final boolean forTask) {
+        final int floor = numberOfActiveTasks / numberOfWorkers;
```
Refactoring similar code to common methods is good and improves readability. 

But during code shuffling and especially if we add functional changes let's not apply also minor stylistic changes. 
We don't usually use `final` for local variables in Connect.

kkonstantine reviewed (commented):
Thanks for the PR @chia7712 

I'd like to return for a detailed review after we exclude stylistic changes that are not necessary. Also, it'd be good to enhance the description besides the example and call out the corner cases for which logic has been added.

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
remove unnecessary 'final'

chia7712 commented:
@kkonstantine thanks for your feedback.

> I'd like to return for a detailed review after we exclude stylistic changes that are not necessary. Also, it'd be good to enhance the description besides the example and call out the corner cases for which logic has been added.

Sorry that I did not follow the code style in connect module. I have reverted all incorrect style. Also, I have updated the description.

kkonstantine reviewed (commented):
Thanks for the quick turnaround @chia7712 !
Much easier to review line my line now (at least for me). 

I'll do my best to review by the end of the week.

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309

chia7712 committed:
Merge branch 'trunk' into KAFKA-12309
